packages/runtime/src/runtimes/react/index.tsx

```diff
@@ -266,31 +262,46 @@ type ElementProps = {
 }
 
 export const Element = memo(
-  forwardRef(function Element({ element }: ElementProps, ref: Ref<unknown>): JSX.Element {
+  forwardRef(function Element(
+    { element }: ElementProps,
+    ref: Ref<ElementImperativeHandle>,
+  ): JSX.Element {
     const elementKey = element.key
     const dispatch = useDispatch()
     const documentKey = useDocumentKey()
-    const [handle, setHandle] = useState<unknown>(null)
+    const imperativeHandleRef = useRef(new ElementImperativeHandle())
+    const findDomNodeCallbackRef = useCallback((current: (() => Element | Text | null) | null) => {
+      imperativeHandleRef.current.callback(() => current?.() ?? null)
+    }, [])
+    const elementCallbackRef = useCallback((current: unknown | null) => {
+      imperativeHandleRef.current.callback(() => current)
+    }, [])
```

Comment:
This is a good question. If the component doesn't forward the ref, then `elementCallbackRef` will never be called. In that case, we rely in `findDOMNode` as a best-effort to get a measurable element so we can place our telegraphs. If the component _does_ forward the ref, then `elementCallbackRef` will be called, but _after_ `findDomNodeCallbackRef` because that component is higher up in the React element tree. So I'm relying on that React behavior here, otherwise they _would_ compete with each other.

I could avoid relying on React's behavior here and instead make it explicit by only using `findDOMNode` is the ref isn't forwarded. I wasn't sure that was possible to detect, but given my explanation above it should be doable if `elementCallbackRef` is called at all.

Will push up a commit to address this 🔜

Reply:
Cool. Makes sense.